src/Microsoft.Kusto.ServiceLayer/DataSource/KustoDataSource.cs

```diff
@@ -484,37 +477,47 @@ public override ScriptFileMarker[] GetSemanticMarkers(ScriptParseInfo parseInfo,
         public override void Refresh()
         {
             // This class caches objects. Throw them away so that the next call will re-query the data source for the objects.
-            databaseMetadata = null;
-            tableMetadata = new ConcurrentDictionary<string, IEnumerable<DataSourceObjectMetadata>>();
-            columnMetadata  = new ConcurrentDictionary<string, IEnumerable<DataSourceObjectMetadata>>();
+            _databaseMetadata = null;
+            _tableMetadata = new ConcurrentDictionary<string, IEnumerable<TableMetadata>>();
+            _columnMetadata  = new ConcurrentDictionary<string, IEnumerable<DataSourceObjectMetadata>>();
+            _folderMetadata = new ConcurrentDictionary<string, IEnumerable<FolderMetadata>>(StringComparer.OrdinalIgnoreCase);
+            _functionMetadata = new ConcurrentDictionary<string, IEnumerable<FunctionMetadata>>();
```

Comment:
A comment here on what contains what would be really helpful. Is this correct? I am sure I missed something:
- _folderMetadata contains folder information for each db. Key is db urn, or parent folder urn. Value is list of folders.
- _tableMetadata contains information for each folder. Key is folder urn. Value is list of tables.
- _columnMetadata contains information for each table. Key is table urn. Value is list of columns.
- _functionMetadata contains information for each folder. Key is folder urn. Value is list of functions.